04-onion-routing.md

```diff
@@ -263,39 +264,78 @@ It is formatted according to the Type-Length-Value format defined in [BOLT #1](0
+    - For the first node in the blinded route:
+      - MUST include the `blinding_point` provided by the recipient
+    - If it is the final node:
+      - MUST include `amt_to_forward` and `outgoing_cltv_value`.
+    - Otherwise:
+      - MUST NOT include `amt_to_forward` and `outgoing_cltv_value`.
+  - For every node outside of a blinded route:
     - MUST include `amt_to_forward` and `outgoing_cltv_value`.
-  - For every non-final node:
-    - MUST include `short_channel_id`
-    - MUST NOT include `payment_data`
-  - For the final node:
-    - MUST NOT include `short_channel_id`
-    - if the recipient provided `payment_secret`:
-      - MUST include `payment_data`
-      - MUST set `payment_secret` to the one provided
-      - MUST set `total_msat` to the total amount it will send
-    - if the recipient provided `payment_metadata`:
-      - MUST include `payment_metadata` with every HTLC
-      - MUST not apply any limits to the size of payment_metadata except the limits implied by the fixed onion size
+    - For every non-final node:
+      - MUST include `short_channel_id`
+      - MUST NOT include `payment_data`
+    - For the final node:
+      - MUST NOT include `short_channel_id`
+      - if the recipient provided `payment_secret`:
+        - MUST include `payment_data`
+        - MUST set `payment_secret` to the one provided
+        - MUST set `total_msat` to the total amount it will send
+      - if the recipient provided `payment_metadata`:
+        - MUST include `payment_metadata` with every HTLC
+        - MUST not apply any limits to the size of `payment_metadata` except the limits implied by the fixed onion size
 
 The reader:
-  - MUST return an error if `amt_to_forward` or `outgoing_cltv_value` are not present.
-  - if it is the final node:
-    - MUST treat `total_msat` as if it were equal to `amt_to_forward` if it
-      is not present.
+
+  - If `blinding_point` is set (either in the payload or the outer message):
+    - MUST return an error if it is set in both the payload and the outer message
+    - MUST return an error if `encrypted_recipient_data` is not present.
+    - MUST return an error if the expiry is greater than `encrypted_recipient_data.payment_constraints.max_cltv_expiry`.
+    - MUST return an error if the amount is below `encrypted_recipient_data.payment_constraints.htlc_minimum_msat`.
+    - MUST return an error if the payment uses a feature not included in `encrypted_recipient_data.payment_constraints.allowed_features`.
+    - If it not the final node:
+      - MUST return an error if `encrypted_recipient_data` does not contain `short_channel_id` or `next_node_id`.
+      - MUST return an error if `encrypted_recipient_data` does not contain `payment_relay`.
+      - If `encrypted_recipient_data.payment_relay` does not satisfy its current settings:
+        - MAY return an error.
+      - Otherwise:
+        - MUST use values from `encrypted_recipient_data.payment_relay` to relay the payment.
```

Comment:
Let's rewrite this to make it explicit:

```
- MUST use values from `encrypted_recipient_data.payment_relay` to calculate `amt_to_forward` and `outgoing_cltv_value` as follows:
   - amt_to_forward = amount_msat - ...
   - outgoing_cltv_value = ctlv_expiry - payment_relay.cltv_expiry_delta
```

This turns out to be how I'm actually implementing it, *and* leaves the rest of the text intact (perhaps with a few references elsewhere to the fact that these values can be implied rather than explicit).

Reply:
Concretely, using standard truncating integer arithmetic:

```
 amt_to_forward = (amount_msat - fee_base_msat) * 1000000 + 1000000 + fee_proportional_millionths - 1) / (1000000 + fee_proportional_millionths)
```

Reply:
Neat thanks for the explicit formula, we had an approximation based on `ceil` but this is better!
Done in https://github.com/lightning/bolts/pull/765/commits/485c2b1a5bc5b7dc1eed919a47a036988cc540f4